ring/ring.go

```diff
@@ -823,7 +835,11 @@ func (r *Ring) shuffleShard(identifier string, size int, lookbackPeriod time.Dur
 				if lookbackPeriod == 0 && instance.ReadOnly {
 					continue
 				}
-
+				// With lookback period >0, read only instances are only included if they have not changed read-only status in the lookback window.
+				// If ReadOnlyUpdatedTimestamp is not set, we include the instance, and extend the shard later.
+				if lookbackPeriod > 0 && instance.ReadOnly && instance.ReadOnlyUpdatedTimestamp > 0 && instance.ReadOnlyUpdatedTimestamp < lookbackUntil {
+					continue
+				}
```

Comment:
These two conditional statements mirror the filtering we do in `filterOutReadOnlyInstances`. It's unfortunate to have that same logic in two places. Is there a way to filter read only instances in just one place?

Reply:
I moved the checks into a function used in both places: https://github.com/grafana/dskit/pull/565/commits/a59aa92fcfcd5291ac00b685820ab2ea52213701